tf/modules/ecs/ec2_capacity_provider/launch_template.tf

```diff
@@ -0,0 +1,82 @@
+resource "aws_launch_template" "launch_template" {
+  name                   = "${var.name}_launch_template"
+  instance_type          = var.instance_type
+  image_id               = var.ami_id == null ? data.aws_ami.ecs_optimized.id : var.ami_id
+  vpc_security_group_ids = var.security_group_ids
+  user_data              = base64encode(local.user_data)
+  update_default_version = true
+  key_name               = var.key_name
+
+  ebs_optimized = var.ebs_size_gb > 0
+
+  dynamic "block_device_mappings" {
+    for_each = var.ebs_size_gb > 0 ? [{}] : []
+
+    content {
+      // The instance volume used by Docker
+      device_name = "/dev/xvdcz"
+
+      ebs {
+        volume_size           = var.ebs_size_gb
+        volume_type           = var.ebs_volume_type
+        delete_on_termination = true
+      }
+    }
+  }
+
+  dynamic "block_device_mappings" {
+    for_each = var.data_size_gb > 0 ? [{}] : []
+
+    content {
+      // The instance volume used by Docker
+      device_name = "/dev/xvdf"
+
+      ebs {
+        volume_size           = var.data_size_gb
+        volume_type           = var.data_volume_type
+        delete_on_termination = true
+      }
+    }
+  }
+
+  dynamic "instance_market_options" {
+    for_each = var.use_spot_purchasing ? [{}] : []
+
+    content {
+      market_type = "spot"
+    }
+  }
+
+  dynamic "network_interfaces" {
+    for_each = var.assign_public_ips ? [{}] : []
+
+    content {
+      associate_public_ip_address = true
+      security_groups             = var.security_group_ids
+    }
+  }
+
+  iam_instance_profile {
+    arn = aws_iam_instance_profile.instance_profile.arn
+  }
+}
+
+data "aws_ami" "ecs_optimized" {
+  owners      = ["amazon"]
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["amzn2-ami-ecs-hvm-*-x86_64-ebs"]
+  }
+}
+
+locals {
```

Comment:
A bit contrived, but it might be nicer in the long term to specify this via cloud-init:

```terraform
data "cloudinit_config" "launch_template_cloudinit" {
  part {
    content_type = "text/x-shellscript"
    content = "echo ECS_CLUSTER=${var.cluster_name}" >> /etc/ecs/ecs.config"
    filename = "configure_ecs_cluster_name.sh"
  }

  part {
    content_type = "text/x-shellscript"
    content  = var.additional_user_data
    filename = "configure_ecs_cluster_name.sh"
  }
}
```

Then a downstream user of this module can decide whether they want to provide a shell script, or one of the other supported cloud-init formats: https://cloudinit.readthedocs.io/en/latest/topics/format.html

Reply:
TIL `data "cloudinit_config"`. Agree that's a better option long term.